packages/flutter/lib/src/cupertino/sheet.dart

```diff
@@ -674,14 +716,25 @@ mixin _CupertinoSheetRouteTransitionMixin<T> on PageRoute<T> {
       linearTransition: linearTransition,
       child: _CupertinoDragGestureDetector<T>(
         enabledCallback: () => enableDrag,
-        onStartPopGesture: () => _startPopGesture<T>(route),
+        onStartPopGesture: () => _startPopGesture<T>(route, topGap),
         child: child,
       ),
     );
   }
 
+  @override
+  bool canTransitionFrom(TransitionRoute<dynamic> previousRoute) {
+    if (this is CupertinoSheetRoute<dynamic>) {
+      return !hasCustomTopGap;
+    }
+    return true;
+  }
+
   @override
   bool canTransitionTo(TransitionRoute<dynamic> nextRoute) {
+    if (this is CupertinoSheetRoute<dynamic>) {
+      return !hasCustomTopGap && nextRoute is _CupertinoSheetRouteTransitionMixin;
+    }
     return nextRoute is _CupertinoSheetRouteTransitionMixin;
```

Comment:
Ideally, I think it would be better if it could be simplified down to
```dart
bool canTransitionTo(TransitionRoute<dynamic> nextRoute) {
   return nextRoute is _CupertinoSheetRouteTransitionMixin && !hasCustomTopGap;
}
```

But at worst it could be 
```dart
bool canTransitionTo(TransitionRoute<dynamic> nextRoute) {
   if (this is CupertinoSheetRoute<dynamic> && hasCustomTopGap) {
      return false;
    }
   return nextRoute is _CupertinoSheetRouteTransitionMixin && !hasCustomTopGap;
}